In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [2]:
import pandas as pd

# 붓꽃 데이터 세트를 로딩합니다.
iris= load_iris()

# iris.data는 Iris데이터 세트에서 feature만으로 된 데이터를 numpy로 가지고 있습니다.
iris_data= iris.data

# iris.target은 붓꽃 데이터 세트에서 레이블(결정값) 데이터를 numpy로 가지고 있습니다.
iris_label= iris.target
print('iris target값: ', iris_label)
print('iris target명: ', iris.target_names)

# 붓꽃 데이터 세트를 자세히 보기 위해 DataFrame 으로 변환합니다.
iris_df = pd.DataFrame( data=iris_data, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df.head(3)

iris target값:  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target명:  ['setosa' 'versicolor' 'virginica']


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   label  
0      0  
1      0  
2      0

In [3]:
# iris_label 레이블 데이터 세트
# test_size=0.2 : 전체데이터셋 중 테스트 데이터 세트의 비율이 0.2
# random_state=11: 난수 발생값

X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.2, random_state=11)

In [4]:
# DecisionTreeClassifier 객체 생성
dt_cls = DecisionTreeClassifier( random_state=11)

#학습 수행
dt_cls.fit(X_train, y_train)


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=11,
            splitter='best')

In [5]:
from sklearn.metrics import accuracy_score

# 학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행
pred= dt_cls.predict(X_test)

# accuracy_score(): 정확도 측정(모델기반 예측레이블과 실제레이블)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9333


In [6]:
from sklearn.datasets import load_iris
iris_data = load_iris()
print(type(iris_data))

<class 'sklearn.utils.Bunch'>


In [7]:
keys= iris_data.keys()
print(' 붓꽃 데이터 셋의 키들\n:', keys)

 붓꽃 데이터 셋의 키들
: dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])


In [8]:
print('\n feature_names의 type: ', type(iris_data.feature_names))

print('\n feature_names의 크기: ', len(iris_data.feature_names))

print('\n iris_data.feature_names:', iris_data.feature_names)

print('\n target_names의 type: ', type(iris_data.target_names))

print('\n target_names의 크기: ', len(iris_data.target_names))


 feature_names의 type:  <class 'list'>

 feature_names의 크기:  4

 iris_data.feature_names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names의 type:  <class 'numpy.ndarray'>

 target_names의 크기:  3


In [9]:
print('\n iris_data.target_names: ', iris_data.target_names)

print('\n iris_data.data의 type: ', type(iris_data.data))

print('\n iris_data.data의 shape: ', iris_data.data.shape)

print('\n iris_data[\'data\']: \n', iris_data['data'])

print('\n target의 type: ', type(iris_data.target))

print('\n target의 크기: ', iris_data.target.shape)

print('\n iris_data.target:\n ', iris_data.target)


 iris_data.target_names:  ['setosa' 'versicolor' 'virginica']

 iris_data.data의 type:  <class 'numpy.ndarray'>

 iris_data.data의 shape:  (150, 4)

 iris_data['data']: 
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.1 1.5 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6

# Module Selection 소개

테스트 데이터 세트를 이용하지 않고 학습 데이터 세트로만 학습하고

예측하면 무엇이 문제인지 살펴보겠습니다. 다음예제는 학습과 예측을 동일한 데이터 세트로 수행한 결과입니다.

In [10]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

iris = load_iris()
dt_clf = DecisionTreeClassifier()
train_data = iris.data
train_label= iris.target
dt_clf.fit(train_data, train_label)

#학습 데이터 세트로 예측 수행
pred = dt_clf.predict(train_data)
print('예측 정확도: ', accuracy_score(train_label, pred))

예측 정확도:  1.0


In [11]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

dt_clf = DecisionTreeClassifier()
iris_data = load_iris()

# 학습데이터셋과 테스트데이터셋 분리
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, 
                                                    iris_data.target, 
                                                    test_size=0.3, 
                                                    random_state=121)
#모델링
dt_clf.fit(X_train, y_train)

pred=dt_clf.predict(X_test)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9556


# KFold

In [12]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris=load_iris()
features= iris.data
label= iris.target
dt_clf = DecisionTreeClassifier(random_state= 156)

# 5개의 폴드 세트로 분리하는 KFold 객체와 폴드 세트별 정확도를 담을 리스트 객체를 생성
kfold= KFold(n_splits=5)
cv_accuracy=[]
print('붓꽃 데이터셋의 크기: ', features.shape[0])

붓꽃 데이터셋의 크기:  150


In [13]:
n_iter=0

# KFold 객체의 split()를 호출하면 폴드별 학습용, 검증용 테스트의 row index를 array로 변환
for train_index, test_index in kfold.split(features):
    
    # kfold.split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    
    #학습 및 예측
    dt_clf.fit(X_train, y_train)
    pred= dt_clf.predict(X_test)
    
    n_iter+=1
    
    #반복할 때 마다 정확도 측정
    accuracy= np.round( accuracy_score(y_test, pred), 4) 
    
    print('X_train.shape[0]: ', X_train.shape[0])
    print('X_test.shape[0]: ', X_test.shape[0])
    train_size=X_train.shape[0]
    test_size= X_test.shape[0]
    print('\n#{0} 교차 검증 정확도: {1}, 학습데이터 크기: {2}, 검증데이터 크기: {3}'.format( \
                                                                      n_iter, accuracy, train_size, test_size))
    
    print('#{0} 검증세트 인덱스: {1}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)
    
# 개별 iteration별 정확도를 합하여 평균 정확도 계산
print('\n## 평균 검증 정확도: ', np.mean(cv_accuracy))

X_train.shape[0]:  120
X_test.shape[0]:  30

#1 교차 검증 정확도: 1.0, 학습데이터 크기: 120, 검증데이터 크기: 30
#1 검증세트 인덱스: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]
X_train.shape[0]:  120
X_test.shape[0]:  30

#2 교차 검증 정확도: 0.9667, 학습데이터 크기: 120, 검증데이터 크기: 30
#2 검증세트 인덱스: [30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]
X_train.shape[0]:  120
X_test.shape[0]:  30

#3 교차 검증 정확도: 0.8667, 학습데이터 크기: 120, 검증데이터 크기: 30
#3 검증세트 인덱스: [60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]
X_train.shape[0]:  120
X_test.shape[0]:  30

#4 교차 검증 정확도: 0.9333, 학습데이터 크기: 120, 검증데이터 크기: 30
#4 검증세트 인덱스: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]
X_train.shape[0]:  120
X_test.shape[0]:  30

#5 교차 검증 정확도: 0.7333, 학습데이터 크기: 120, 검증데이터 크기: 30
#5 검증세트 인덱스: [120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 

# Stratified K-Fold

불균형한(imbalanced) 분포도를 가진 레이블(결정클래스) 데이터 집합을 위한 K폴드 방식이다.

불균형한 분포도를 가진 레이블 데이터 집합은 특정 레이블 값이 특이하게 

많거나 매우 적어서 값의 분포가 한쪽으로 치우치는 것을 말한다.


KFold가 레이블 데이터 집합이 원본 데이터 집합의 레이블 분포를 학습 & 테스트에 제대로 분배하지 못하는 경우의 문제를 해결.

-원본데이터의 레이블 분포를 먼저 고려<br>

-이 분포와 동일하게 학습과 검증 데이터 세트를 분배

-KFold가 어떤 문제를 가지고 있는지 확인

-sklearn의 StratifiedKFold 클래스를 이용해 개선

-붓꽃 데이터 세트를 DataFrame으로 생성하고 레이블값의 분포도를 확인

In [14]:
from sklearn.datasets import load_iris
import pandas as pd

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold

iris= load_iris()
iris_df = pd.DataFrame( data= iris.data, columns=iris.feature_names )
iris_df.head(10)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2
5                5.4               3.9                1.7               0.4
6                4.6               3.4                1.4               0.3
7                5.0               3.4                1.5               0.2
8                4.4               2.9                1.4               0.2
9                4.9               3.1                1.5               0.1

In [15]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [16]:
iris_df['label'] = iris.target

In [17]:
iris_df['label'].value_counts()

2    50
1    50
0    50
Name: label, dtype: int64

In [18]:
kfold=KFold(n_splits=3)
n_iter=0

for train_index, test_index in kfold.split(iris_df):
    n_iter+=1
    label_train = iris_df['label'].iloc[train_index]
    label_test = iris_df['label'].iloc[test_index]
    print('\n\n## 교차 검증 {}'.format(n_iter))
    print('학습 레이블 데이터 분포: \n', label_train.value_counts())
    print('\n검증 레이블 데이터 분포: \n', label_test.value_counts())



## 교차 검증 1
학습 레이블 데이터 분포: 
 2    50
1    50
Name: label, dtype: int64

검증 레이블 데이터 분포: 
 0    50
Name: label, dtype: int64


## 교차 검증 2
학습 레이블 데이터 분포: 
 2    50
0    50
Name: label, dtype: int64

검증 레이블 데이터 분포: 
 1    50
Name: label, dtype: int64


## 교차 검증 3
학습 레이블 데이터 분포: 
 1    50
0    50
Name: label, dtype: int64

검증 레이블 데이터 분포: 
 2    50
Name: label, dtype: int64



StratifiedKFold는 이렇게 KFold를 분할된 레이블 데이터 세트가 

전체 레이블값의 분포도를 반영하지 못하는 문제를 해결해줍니다.

동일한 데이터 분할을 StratifiedKFold로 수행하고 학습 검증 레이블 데이터의 분포도를 확인.

- KFold와 StratifiedKFold 와의 차이점?

- => StratifiedKFold는 레이블 데이터 분포도에 따라 학습/검증 데이터를 나누기 때문에

- split() 메서드에 인자로 피쳐데이터 세트뿐만 아니라 레이블 데이터 세트도 반드시 필요합니다.


In [19]:
# KFold fold세트: 3
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=3)
n_iter=0

for train_index, test_index in skf.split(iris_df, iris_df['label']): # feature_dataset:iris_df, lable_dataset:iris_df['label']
    n_iter+=1
    label_train = iris_df['label'].iloc[train_index]
    label_test = iris_df['label'].iloc[test_index]
    #     print('label_train\n', label_train)
    #     print('\nlabel_test\n', label_test,'\n')

    print('## 교차 검증 {}'.format(n_iter))
    print('학습 레이블 데이터 분포\n', label_train.value_counts())
    print('검증 레이블 데이터 분포\n', label_test.value_counts())

## 교차 검증 1
학습 레이블 데이터 분포
 2    33
1    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포
 2    17
1    17
0    17
Name: label, dtype: int64
## 교차 검증 2
학습 레이블 데이터 분포
 2    33
1    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포
 2    17
1    17
0    17
Name: label, dtype: int64
## 교차 검증 3
학습 레이블 데이터 분포
 2    34
1    34
0    34
Name: label, dtype: int64
검증 레이블 데이터 분포
 2    16
1    16
0    16
Name: label, dtype: int64


학습레이블과 검증 레이블의 데이터 값의 분포도가 동일하게 할당되어있다.

동일하게 할당해야 레이블 0,1,2 모두를 학습할 수 있다. 이에 기반해 검증을 수행할 수 있다.

In [20]:
dt_clf = DecisionTreeClassifier(random_state=156)

skfold= StratifiedKFold(n_splits=3)

n_iter=0
cv_accuracy=[]

# StratifiedKFold의 split() 호출시 반드시 레이블 데이터 세트도 추가 입력 필요
for train_index, test_index in skfold.split(features, label):
    # split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    
    #학습 및 예측
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)
    
    # 반복시마다 정확도 측정
    n_iter+=1
    accuracy= np.round(accuracy_score(y_test, pred), 4)
    train_size=X_train.shape[0]
    test_size= X_test.shape[0]
    print('\n#{} 교차 검증 정확도 {}\n학습 데이터크기: {}\n검증 데이터크기: {}'.format(
        n_iter, 
        accuracy, 
        train_size, 
        test_size))
    print('#{} 검증세트 인덱스 : {}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)
    
    
# 교차 검증별 정확도 및 평균 정확도 계산
print('\n\n# 교차 검증별 정확도: ', np.round(cv_accuracy,4))
print('## 평균 검증 정확도: ', np.mean(cv_accuracy))
    
    


#1 교차 검증 정확도 0.9804
학습 데이터크기: 99
검증 데이터크기: 51
#1 검증세트 인덱스 : [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116]

#2 교차 검증 정확도 0.9216
학습 데이터크기: 99
검증 데이터크기: 51
#2 검증세트 인덱스 : [ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133]

#3 교차 검증 정확도 0.9792
학습 데이터크기: 102
검증 데이터크기: 48
#3 검증세트 인덱스 : [ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  84  85
  86  87  88  89  90  91  92  93  94  95  96  97  98  99 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]


# 교차 검증별 정확도:  [0.9804 0.9216 0.9792]
## 평균 검증 정확도:  0.9604


분류(Classification)에서 교차 검증은 KFold가 아니라 StratifiedKFold로 분할돼야한다.

회귀(Regression)에서는 StratifiedKFold가 지원되지 않는다.

회귀의 결정값은 인산값 형태의 레이블이 아니라 연속된 숫자이기 때문에

결정값별 분포를 정하는 의미가 없다.


## 교차 검증을 보다 간편하게 -cross_val_score()

사이킷런에서 제공하는 교차 검증을 좀 더 편리하게 수행할 수 있게하는 api

[KFold]
1. 폴드세트를 설정한다.
2. for loop에서 반복으로 학습 및 테스트데이터의 인덱스를 추출
3. 반복적으로 학습과 예측을 수행하고 예측 성능을 반환

cross_val_score() 은 위의 일련의 과정을 한꺼번에 수행해주는 api이다.

**cross_val_score(**<BR>
    estimator,<BR>
    X,<BR>
    y=None,<br>
    scoring=None,<br>
    cv=None,<br>
    n_jobs=1,<br>
    verbose=0,<br>
    fit_params=None,<br>
    pre_dispatch='2*n_jobs'<br>
) <br>

X: feature dataset<br>
y: label dataset<br>
scoring: 예측 성능 평가 지표를 기술<br>
cv: 교차 검증 폴드 수<br>

cross_val_score() 수행후 반환값은 scoring 파라미터로 지정된 성능 지표 측정값을 배열형태로 반환

cross_val_score()는 classifier가 입력되면 StratifiedKFold 방식으로 레이블값의 분포에 따라

학습/ 테스트 세트를 분할.

In [21]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data = load_iris()
dt_clf = DecisionTreeClassifier(random_state=156)

data= iris_data.data
label= iris_data.target

# 성능 지표 정확도 (accuracy), 교차 검증 세트는 3개
scores= cross_val_score( dt_clf, data, label, scoring='accuracy', cv=3)
print('교차 검증별 정확도: ', np.round(scores,4)) # np.round( value, 보여줄 소수점아래자리수)==> 소수점 4째자리까지 보여줌.
print('평균 검증 정확도  : ', np.round(np.mean(scores),4 ))

교차 검증별 정확도:  [0.9804 0.9216 0.9792]
평균 검증 정확도  :  0.9604


cross_val_score()는 cv로 지정된 횟수 만큼 scoring 파라미터로 지정된 평가 지표로 평가 결괏값을 배열로 반환한다.

일반적으로 이를 평균해 평가 수치로 사용한다.

cross_val_score() API는 내부에서 Estimator를 학습(fit), 예측(predict), 평가(evaluation) 시켜주므로

간단하게 교차검증을 수행할 수 있다.

cross_val_score() 수행결과와 StraitifiedKFold 수행결과를 비교해보면

각 교차 검증별 정확도와 평균검증 정확도가 모두 동일함을 알 수있다.

이는 **cross_val_score()가 내부적으로 StratifiedKFold를 이용하기 때문이다.**

**cross_val_score()**는 **단 하나의 평가 지표**만을 가능하지만 

**cross_validate()**는 **여러개의 평가 지표를 반환**할 수있다.

또한 학습 데이터에 대한 성능 평가 지표와 수행 시간도 같이 제공한다.

cross_val_score() 하나로도 대부분의 경우 쉽게 사용하므로 cross_validate()에 대한 예제는 건너뛴다.

# GridSearchCV : 교차 검증과 최적 하이퍼파라미터 튜닝을 한번에 처리.

하이퍼 파라미터 : 머신러닝 알고리즘을 구성하는주요 구성요소, 하이퍼파라미터 값을 조정해 알고맂므의 예측 성능을 개선.

GridSearchCV API를 이용해 Classifier나 Regressor과 같은 알고리즘에 사용되는 

하이퍼 파라미터를 순차적으로 입력하면서 편리하게 최적의 파라미터를 도출할 수있는 방안을 제공.

결정트리 알고리즘의 여러 하이퍼 파라미터를 순차적으로 변경하면서 최고의 성능을 가지는 파라미터 조합을

찾고자 한다면 다음과 같이 파라미터의 집하을 만들고 이를 순차적으로 적용하면서 최적화를 수행할 수있다.

grid_parameters={ 'max_depth': [1,2,3],<br></br>
                  'min_samples_split':[2,3]<br>
                }<br>
                
                
하이퍼 파라미터는 순차적으로 적용되며, 총 6회에 걸쳐 파라미터를 순차적으로 바꿔 실행하면서 

최적의 파라미터와 수행결과를 도출한다.

for루프로 모든 파라미터를 번갈아 입력하면서 학습시키는 방법을 api 레벨에서 제공.

GridSearchCV는 교차 검증을 기반으로 이 하이퍼 파라미터의 최적 값을 찾게 해준다.

즉, 데이터 세트를 cross-validation을 위한 학습/테스트 세트로 자동으로 분할한 뒤 

하이퍼 파라미터 그리드에 기술된 모든 파라미터를 순차적으로 적용해 최적의 파라미터를 찾을 수 있게 해준다.


GridSearchCV는 사용자가 튜닝하고자 하는 여러 종류의 하이퍼 파라미터를 다양하게 테스트하면서

최적의 파라미터를 편리하게 찾게 해주지만 동시에 순차적으로 파라미터를 테스트하므로

수행시간이 상대적으로 오래걸린다.

6회에 걸쳐 하이퍼 파라미터를 변경하면서 교차 검증 데이터 세트에 수행 성능을 측정.

CV: 3회 ==> 개별 파라미터 조합마다 3개의 폴딩세트를 3회에 걸쳐 학습/평가해서 평균값으로 성능을 측정

6개의 파라미터 조합 => 총 CV 3회 * 6개 파라미터 조합 = 18회의 학습/평가가 이뤄진다.

- estimator: classifier, regressor, pipeline


- param_grid: key + 리스트 값을 가지는 딕셔너리가 주어진다. estimator의 튜닝을 위해 파라미터명과 사용될 여러 파라미터 값을 지정.


- scoring: 예측 성능을 측정할 평가방법을 지정한다.<br>
<t> 보통은 sklearn 성능 평가 지표를 지정하는 문자열(ex:정확도)로 지정하나 별도의 성능 평가 지표 함수도 지정
    
    
- cv: 교차 검증을 위해 분할되는 학습/테스트 세트의 개수를 지정.


- refit: default가 True이며, True로 생성시 가장 최적의 하이퍼 파라미터를 찾은 뒤 <br>
<t> 입력된 estimator 객체를 해당 하이퍼파라미터로 재학습
    
    
    
결정트리 알고리즘이 여러가지 최적화 파라미터를 순차적으로 적용해 붓꽃 데이터를 예측 분석하는데 GridSearchCV를 이용.

1. train_test_split()을 이용해 학습데이터와 테스트 데이터를 먼저 분리


2. 학습 데이터에서 GridSearchCV를 이용해 최적 하이퍼 파라미터를 추출


3. 결정트리 알고리즘을 구현한 DecisionTreeClassifier의 중요 하이퍼파라미터은<br>
<t> max_depth, min_samples_split의 값을 변화 시키면서 최적화를 진행


4. 테스트할 하이퍼 파라미터 세트는 딕셔너리 형태로 <br>
<t> 하이퍼파라미터의 명칭은 문자열 Key값으로, 하이퍼 파라미터 값은 리스트형으로 설정.

In [22]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# 데이터를 로딩하고 학습데이터와 테스트 데이터로 분리
iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size=0.2, random_state=121)

dtree = DecisionTreeClassifier()

# 파라미터를 딕셔너리 형태로 설정
parameters = {'max_depth': [1,2,3], 'min_samples_split': [2,3]}


In [23]:
import pandas as pd

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정.
## refit= True가 default이다.
## refit=True => 가장 좋은 파라미터 설정으로 재학습 시킴.

grid_dtree = GridSearchCV( dtree, param_grid=parameters, cv=3, refit=True ) #kfold횟수 3번

# 붓꽃 학습데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과를 추출해 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[ ['params', 'mean_test_score', 'rank_test_score', 
            'split0_test_score', 'split1_test_score', 'split2_test_score']] #3번의 kfold

C:\Users\USER\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:122: FutureWarning: You are accessing a training score ('split0_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
C:\Users\USER\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:122: FutureWarning: You are accessing a training score ('split1_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
C:\Users\USER\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:122: FutureWarning: You are accessing a training score ('split2_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
C:\Users\USER\Anaconda3\lib\site-packages\sklearn\utils\de

params  mean_test_score  rank_test_score  \
0  {'max_depth': 1, 'min_samples_split': 2}         0.700000                5   
1  {'max_depth': 1, 'min_samples_split': 3}         0.700000                5   
2  {'max_depth': 2, 'min_samples_split': 2}         0.958333                3   
3  {'max_depth': 2, 'min_samples_split': 3}         0.958333                3   
4  {'max_depth': 3, 'min_samples_split': 2}         0.966667                1   
5  {'max_depth': 3, 'min_samples_split': 3}         0.966667                1   

   split0_test_score  split1_test_score  split2_test_score  
0              0.700                0.7               0.70  
1              0.700                0.7               0.70  
2              0.925                1.0               0.95  
3              0.925                1.0               0.95  
4              0.950                1.0               0.95  
5              0.950                1.0               0.95

주요 컬럼별 의미는 다음과 같이 정리할 수 있다.

- params 칼럼에는 수행할 때마다 적용된 개별 하이퍼 파라미터값을 나타낸다.


- rank_test_score는 하이퍼 파라미터별로 성능이 좋은 score 순위를 나타낸다.<br>
<t> 1이 가장 뛰어난 순위이며 이때 파라미터가 최적의 하이퍼 파라미터이다.


- mean_test_score는 개별 하이퍼 파라미터별로 CV의 폴딩 테스트 세트에 대해 총 수행한 평가 평균값이다.<br>

GridSearchCV 객체의 fit()을 수행하면 최고 성능을 나타낸 하이퍼 파라미터의 값과

그때의 평가 결과값이 각각 best_params_, best_score_ 속성에 기록된다.

즉, cv_results_의 rank_test_score가 1일 때의 값이다.

이 속성을 이용해 최적 하이퍼 파라미터의 값과 그때의 정확도를 알아본다.

In [24]:
print('GridSearchCV 최적 파라미터: ', grid_dtree.best_params_ )
print('GridSearchCV 최고 정확도  : {:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터:  {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도  : 0.9667


max_depth: 3일때<br>
min_samples_split:2 일때 

검증용 폴드 세트에서 평균 최고 정확도가 96.67% 로 측정됐다.

GridSearchCV 객체의 생성 파라미터로 refit=True가 Default이다.

refit=True이면 GridSearchCV가 최적 성능을 나타내는 하이퍼파라미터로 

Estimator를 학습해 best_estimator_로 저장한다.

이미 학습된 best_estimator_을 이용해 

앞에서 train_test_split()으로 분리한 테스트 데이터 세트에 대해 예측하고 성능을 평가

In [25]:
# GridSearchCV 의 refit()으로 이미 학습된 estimator 반환
estimator = grid_dtree.best_estimator_

# GridSearchCV의 best_estimator_는 이미 최적 학습이 됐으므로 별도 학습이 필요 없음.
pred = estimator.predict(X_test)
print('테스트 데이터 세트 정확도: {:.4f}'.format(accuracy_score(y_test, pred)))

테스트 데이터 세트 정확도: 0.9667


일반적으로 학습데이터를 GridSearchCV를 이용해 최적 하이퍼 파라미터 튜닝을 수행한뒤에 별도의 테스트 세트에서

이를 평가하는 것이 일반적인 머신러닝 모델 적용방법이다.